core/embed/rust/src/trezorhal/wordlist.rs

```diff
@@ -0,0 +1,161 @@
+use super::ffi;
+use core::cmp::Ordering;
+use cstr_core::CStr;
+
+/// Holds all the possible words with the possibility to interact
+/// with the "list" - filtering it further, getting their count, etc.
+pub struct Wordlist(&'static [*const cty::c_char]);
+
+impl Wordlist {
+    /// Initialize BIP39 wordlist.
+    pub fn bip39() -> Self {
+        Self(unsafe { &ffi::BIP39_WORDLIST_ENGLISH })
+    }
+
+    /// Initialize SLIP39 wordlist.
+    pub fn slip39() -> Self {
+        Self(unsafe { &ffi::SLIP39_WORDLIST })
+    }
+
+    /// Only leaves words that have a specified prefix. Throw away others.
+    pub fn filter_prefix(&self, prefix: &str) -> Self {
+        let mut start = 0usize;
+        let mut end = self.0.len();
+        for (i, word) in self.0.iter().enumerate() {
+            // SAFETY: We assume our slice is an array of 0-terminated strings.
+            match unsafe { prefix_cmp(prefix, *word) } {
+                Ordering::Less => {
+                    start = i + 1;
+                }
+                Ordering::Greater => {
+                    end = i;
+                    break;
+                }
+                _ => {}
+            }
+        }
+        Self(&self.0[start..end])
+    }
+
+    /// Get a word at the certain position.
+    pub fn get(&self, index: usize) -> Option<&'static str> {
+        // SAFETY: we assume every word in the wordlist is a valid 0-terminated UTF-8
+        // string.
+        self.0
+            .get(index)
+            .map(|word| unsafe { from_utf8_unchecked(*word) })
+    }
+
+    /// How many words are currently in the list.
+    pub const fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Iterator of all current words.
+    pub fn iter(&self) -> impl Iterator<Item = &'static str> {
+        self.0
+            .iter()
+            .map(|word| unsafe { from_utf8_unchecked(*word) })
+    }
+}
+
+/// Compare word from wordlist to a prefix.
+///
+/// The comparison returns Less if the word comes lexicographically before all
+/// possible words starting with `prefix`, and Greater if it comes after.
+/// Equal is returned if the word starts with `prefix`.
+unsafe fn prefix_cmp(prefix: &str, word: *const cty::c_char) -> Ordering {
+    // SAFETY: we assume `word` is a pointer to a 0-terminated string.
+    for (i, prefix_char) in prefix.as_bytes().iter().enumerate() {
+        let word_char = unsafe { *(word.add(i)) } as u8;
+        if word_char == 0 {
+            // Prefix is longer than word.
+            return Ordering::Less;
+        } else if *prefix_char != word_char {
+            return word_char.cmp(prefix_char);
+        }
+    }
+    Ordering::Equal
+}
+
+unsafe fn from_utf8_unchecked<'a>(word: *const cty::c_char) -> &'a str {
```

Comment:
Unsafe functions should have a docstring documenting their safety assumptions. This function can be made to return any lifetime which should probably be also documented.